allennlp/tests/training/metrics/bleu_test.py

```diff
@@ -0,0 +1,110 @@
+# pylint: disable=protected-access,not-callable
+
+from collections import Counter
+import math
+
+import numpy as np
+import torch
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.training.metrics import BLEU
+
+
+class BleuTest(AllenNlpTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.metric = BLEU(ngram_weights=(0.5, 0.5))
+
+    def test_get_valid_tokens_mask(self):
+        tensor = torch.tensor([[1, 2, 3, 0],
+                               [0, 1, 1, 0]])
+        result = self.metric._get_valid_tokens_mask(tensor, set((0,)))
+        result = result.long().numpy()
+        check = np.array([[1, 1, 1, 0],
+                          [0, 1, 1, 0]])
+        np.testing.assert_array_equal(result, check)
+
+    def test_ngrams(self):
+        tensor = torch.tensor([1, 2, 3, 1, 2, 0])
+
+        # Unigrams.
+        counts = Counter(self.metric._ngrams(tensor, 1, set((0,))))
+        unigram_check = {(1,): 2, (2,): 2, (3,): 1}
+        assert counts == unigram_check
```

Comment:
I'm curious about the use of bare assert statements inside of essentially a `unittest.TestCase` class?

```python
import unittest


class Test(unittest.TestCase):
    def test_bare(self):
        a = [1, 2, 3]
        b = [3, 4, 5]
        assert a == b

    def test_self_assert(self):
        a = [1, 2, 3]
        b = [3, 4, 5]
        self.assertEqual(a, b)


if __name__ == '__main__':
    unittest.main()

```

The output is much more descriptive by using (in this particular case) `self.assertEqual`?

```
FF
======================================================================
FAIL: test_bare (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "", line 8, in test_bare
    assert a == b
AssertionError

======================================================================
FAIL: test_self_assert (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "", line 13, in test_self_assert
    self.assertEqual(a, b)
AssertionError: Lists differ: [1, 2, 3] != [3, 4, 5]

First differing element 0:
1
3

- [1, 2, 3]
+ [3, 4, 5]

----------------------------------------------------------------------
Ran 2 tests in 0.001s

FAILED (failures=2)
```

Reply:
Given that we're using pytest, I think the output from a raw assert failure is actually just as, if not more, descriptive:

Here's an `assertDictEqual` failure:
![assertdictequal](https://user-images.githubusercontent.com/8812459/47880167-bfcb0900-dddf-11e8-8b9b-a67807f2cffc.png)
Here's a bare assert failure:
![base_assert](https://user-images.githubusercontent.com/8812459/47880139-a75aee80-dddf-11e8-95eb-9d9aab12db4b.png)

Reply:
@epwalsh 

I'm curious why you would still write tests using the old unittest style then? What's the advantage?

Reply:
I agree with you, @epwalsh. Bare assert seems more descriptive. It's also great to just compose raw Python instead of use custom functions.